scripts/04-effect-for-beginners.md

```diff
@@ -0,0 +1,1622 @@
+
+---
+
+A simple `Effect`
+```ts
+type Simple = Effect<never, Error, number>;
+```
+
+notes:
+Lets break down this effect. Like the unit effect, its requirements field is never, meaning it does not have any dependencies. Unlike the unit effect, its Error type is Error, this means that this Effect may fail with an Error object. The final value type is number, meaning this Effect will yield a number if it succeeds.
+
+---
+
+Creating Effects
+```ts
+// type: Effect<never, never, number>
+const success = Effect.succeed(42);
+
+// type: Effect<never, Error, never>
+const failure = Effect.fail(new Error())
+```
+
+notes:
+To create an Effect from a value we can use the succeed and fail constructor functions. Succeed takes a value and returns an Effect that can only ever succeed with that value. Along the same lines, fail takes a value and returns an Effect that can only ever fail with that value.
+
+---
+Effects compose together
+```ts
+function divide(x: number, y: number): Effect.Effect<never, Error, number> {
+  if (y === 0) {
+    return Effect.fail(new Error("divide by zero"));
+  }
+  return Effect.succeed(x / y);
+}
+```
+
+notes:
+The power of the Effect type comes from its ability to compose together with other Effects. Here is a basic division function. First it checks if the divisor is 0, and if so returns a effect failure containing an Error. Else it returns a effect success containing the result of the division.
+
+In the return type is where the magic happens. See how both the success case and the failure case merge into a single type that describes both outcomes. It is now abundantly clear that a result of this divide function either returns a number or an Error.
+
+---
+
+Delaying Computations
+
+notes:
+Remember how an Effect represents the description of some behavior to be ran at a later point. Well, when we use the succeed and fail constructors, the value passed to them is evaluated when the effect is constructed, not when it is run. 
+
+However, when working with code that has side effects, it might be important to defer those side effects until the effect is run.
+
+---
+
+A synchronous computations
+```ts
+// type: Effect<never, never, number>
+const program = Effect.sync(() => {
+  console.log("Hello, World!") // side effect
+  return 42 // return value
+})
+```
+
+notes:
+To do this we can use the sync constructor. Sync returns an Effect with a value of the return type of the passed function, but this function is only ran when the Effect itself is ran.
+
+Here this Effect represents a program that when run will log to the console and return a number.
+
+Its important to note that sync always returns an Effect with a error of type never, this is because the function passed to it should be guaranteed to never throw.
+
+---
+A synchronous computations that could throw
+```ts
+// type: Effect<never, unknown, any>
+const program = Effect.try(
+  () => JSON.parse("")
+)
+```
+
```

Comment:
worth showing the variant where you can directly map the error to a known type, having `unknown` in the error channel isn't advisable